core/authority-discovery/primitives/src/lib.rs

```diff
@@ -14,18 +14,19 @@
 // You should have received a copy of the GNU General Public License
 // along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
 
-//! Common consensus primitives.
+//! Runtime Api to help discover authorities.
 
 #![cfg_attr(not(feature = "std"), no_std)]
 
-use codec::Codec;
 use client::decl_runtime_apis;
+use codec::Codec;
 use rstd::vec::Vec;
 
 decl_runtime_apis! {
-	/// Common consensus runtime api.
-	pub trait ConsensusApi<AuthorityId: Codec> {
-		/// Returns the set of authorities of the currently active consensus mechanism.
-		fn authorities() -> Vec<AuthorityId>;
-	}
+      pub trait AuthorityDiscoveryApi<AuthorityId: Codec> {
+            fn public_key() -> Option<AuthorityId>;
+            fn authorities() -> Vec<AuthorityId>;
+            fn sign(payload: Vec<u8>) -> Option<Vec<u8>>;
```

Comment:
Note that the calls to `public_key` and `sign` can race against a local session key changing. I'd recommend making it a single atomic operation because of this. We probably need the runtime to construct the payload because of that.

Reply:
There are two possible races, depending on whether we first retrieve the public key or whether we first sign the payload.

### First signing then retrieving public key

In that case the *key*, *signature* validation by other nodes would fail, but given that we put our external addresses on the Dht periodically this should not lead to any failures.

### First retrieving public key then signing

Given that the public key is changed, I am assuming that the public key is no longer needed, thereby other nodes wouldn't query the Dht for the public key anymore. While this results in an invalid dht entry, that entry would never be read by anyone.

---

I am guessing that making this an atomic operation does not only involve being atomic within the runtime api but also all the way down to the key store. 

How often do we estimate a session key to change? With the above failure scenarios in mind, does it need to block this feature?

Reply:
As discussed offline, note added in https://github.com/paritytech/substrate/pull/3385/commits/e8de9c65cdbb491431847bff345ae772f182cc82.